pkg/gatewaysecret/handler_test.go

```diff
@@ -0,0 +1,351 @@
+	tests := []struct {
+		name    string
+		secret  *apicorev1.Secret
+		want    time.Time
+		wantErr bool
+	}{
+		{
+			name: "valid lastModifiedAt annotation",
+			secret: &apicorev1.Secret{
+				ObjectMeta: apimetav1.ObjectMeta{
+					Annotations: map[string]string{
+						"lastModifiedAt": "2024-11-01T00:00:00Z",
+					},
+				},
+			},
+			want:    time.Date(2024, 11, 1, 0, 0, 0, 0, time.UTC),
+			wantErr: false,
+		},
+		{
+			name: "missing lastModifiedAt annotation",
+			secret: &apicorev1.Secret{
+				ObjectMeta: apimetav1.ObjectMeta{
+					Annotations: map[string]string{},
+				},
+			},
+			want:    time.Time{},
+			wantErr: true,
+		},
+		{
+			name: "invalid lastModifiedAt annotation key",
+			secret: &apicorev1.Secret{
+				ObjectMeta: apimetav1.ObjectMeta{
+					Annotations: map[string]string{
+						"LastModifiedAt": "2024-11-01T00:00:00Z",
+					},
+				},
+			},
+			want:    time.Time{},
+			wantErr: true,
+		},
+		{
+			name: "invalid lastModifiedAt annotation time format",
+			secret: &apicorev1.Secret{
+				ObjectMeta: apimetav1.ObjectMeta{
+					Annotations: map[string]string{
+						"lastModifiedAt": "2024-11-01T00:00:00",
+					},
+				},
+			},
+			want:    time.Time{},
+			wantErr: true,
+		},
+	}
+	for _, testcase := range tests {
+		t.Run(testcase.name, func(t *testing.T) {
+			got, err := gatewaysecret.GetValidLastModifiedAt(testcase.secret)
+			if (err != nil) != testcase.wantErr {
+				t.Errorf("GetValidLastModifiedAt() error = %v, wantErr %v", err, testcase.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, testcase.want) {
+				t.Errorf("GetValidLastModifiedAt() got = %v, want %v", got, testcase.want)
+			}
+		})
+	}
+}
+
+func TestGatewaySecretRequiresUpdate(t *testing.T) {
+	type args struct {
+		gwSecret *apicorev1.Secret
+		caCert   certmanagerv1.Certificate
+	}
+	tests := []struct {
+		name string
+		args args
+		want bool
+	}{
+		{
+			name: "gateway secret is newer than CA certificate",
+			args: args{
+				gwSecret: &apicorev1.Secret{
+					ObjectMeta: apimetav1.ObjectMeta{
+						Annotations: map[string]string{
+							"lastModifiedAt": "2024-11-01T00:00:10Z",
+						},
+					},
+				},
+				caCert: certmanagerv1.Certificate{
+					Status: certmanagerv1.CertificateStatus{
+						NotBefore: &apimetav1.Time{
+							Time: time.Date(2024, 11, 1, 0, 0, 5, 0, time.UTC),
+						},
+					},
+				},
+			},
+			want: false,
+		},
+		{
+			name: "gateway secret is older than CA certificate",
+			args: args{
+				gwSecret: &apicorev1.Secret{
+					ObjectMeta: apimetav1.ObjectMeta{
+						Annotations: map[string]string{
+							"lastModifiedAt": "2024-11-01T00:00:00Z",
+						},
+					},
+				},
+				caCert: certmanagerv1.Certificate{
+					Status: certmanagerv1.CertificateStatus{
+						NotBefore: &apimetav1.Time{
+							Time: time.Date(2024, 11, 1, 0, 0, 5, 0, time.UTC),
+						},
+					},
+				},
+			},
+			want: true,
+		},
+		{
+			name: "gateway secret is newer but has parsing error for lastModifiedAt",
```

Comment:
When there is a parsing error, we don't know if it's newer. Maybe `gateway secret has parsing error for lastModifiedAt` is enough. But that is a very important test case, it documents that when there is a problem with `lastModifiedAt`, we consider this as "requires update", which makes sense, as the update should "fix" the invalid condition.